{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "d2ee187f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bycon import *\n",
    "from pymongo import MongoClient\n",
    "\n",
    "# Connect to MongoDB\n",
    "client = MongoClient()\n",
    "db = client.progenetix\n",
    "bs = db.biosamples\n",
    "\n",
    "# Read the data frame\n",
    "df = pd.read_csv(\"../temp/mapfile.tsv\", sep='\\t', header=0, low_memory = False)\n",
    "df = df[0:1000]\n",
    "# Create legacy ids / external references\n",
    "df['case_id'] = 'pgx:TCGA-' + df['case_id']\n",
    "df['sample_id'] = 'pgx:TCGA-' + df['sample_id']\n",
    "\n",
    "# Create placeholder for variant id, gets created while import\n",
    "df['variant_id'] = [''] * len(df) # could use VARIANT_CLASS\n",
    "\n",
    "# alternate_bases_1 is the same as reference_bases\n",
    "df['alternate_bases'] = df['alternate_bases_2']\n",
    "\n",
    "# Naming convention from pgx\n",
    "df['variant_state_id'] = ['SO:0001059'] * len(df)\n",
    "df['reference_name'] = df['chromosome'].str.slice(start=3)\n",
    "df['variant_types'] = df['variant_type']\n",
    "\n",
    "# Adding sequence ontologies - http://www.sequenceontology.org/browser/\n",
    "df.loc[df['variant_type'] == 'SNP', 'specific_so'] = 'SO:0001483'\n",
    "df.loc[df['variant_type'] == 'TNP', 'specific_so'] = 'SO:0002007'\n",
    "df.loc[df['variant_type'] == 'ONP', 'specific_so'] = 'SO:0002007'\n",
    "df.loc[df['variant_type'] == 'DEL', 'specific_so'] = 'SO:0000159'\n",
    "df.loc[df['variant_type'] == 'INS', 'specific_so'] = 'SO:0000667'\n",
    "\n",
    "# Convert 1-based MAF files to 0-based\n",
    "df.loc[df['variant_type'] == 'SNP', 'start'] -= 1\n",
    "df.loc[df['variant_type'] == 'TNP', 'start'] -= 1\n",
    "df.loc[df['variant_type'] == 'ONP', 'start'] -= 1\n",
    "df.loc[df['variant_type'] == 'DEL', 'start'] -= 1\n",
    "df.loc[df['variant_type'] == 'INS', 'end'] -= 1 # explanation at https://www.biostars.org/p/84686/\n",
    "\n",
    "# Generate callset_ids per aliquot for import and map sample_id to biosample_id\n",
    "biosample_mapping = {}\n",
    "for aliquot in set(df['aliquot_id']):\n",
    "    cs_id = generate_id('pgxcs')\n",
    "    df.loc[df['aliquot_id'] == aliquot, 'callset_id'] = cs_id\n",
    "\n",
    "    sample_id = df.loc[df['aliquot_id'] == aliquot, 'sample_id'].iloc[0]\n",
    "    if sample_id not in biosample_mapping:\n",
    "        hit = bs.find({\"external_references.id\": {'$regex': sample_id}, \"biosample_status.id\": \"EFO:0009656\"})\n",
    "        for entry in hit:\n",
    "            biosample_mapping[sample_id] = (entry['id'], entry['individual_id'])\n",
    "            break\n",
    "    biosample_id, individual_id = biosample_mapping.get(sample_id, ('', ''))\n",
    "\n",
    "    df.loc[df['aliquot_id'] == aliquot, 'biosample_id'] = biosample_id\n",
    "    df.loc[df['aliquot_id'] == aliquot, 'individual_id'] = individual_id\n",
    "\n",
    "\n",
    "df['variant_types'] = df['variant_type']\n",
    "# Clean up\n",
    "df = df[['biosample_id', 'variant_id', 'callset_id', 'reference_name', 'start', 'end',\n",
    "        'reference_bases', 'alternate_bases', 'hgvsc', 'variant_classification', 'variant_state_id', \n",
    "        'hgvsp', 'hgvsp_short', 'aliquot_id', 'reference_id', 'case_id', 'sample_id', 'variant_types',\n",
    "        'specific_so']]\n",
    "\n",
    "# variants_in_db = df.dropna(subset = ['biosample_id'])\n",
    "# new = df[df['biosample_id'].isna()]\n",
    "test = df\n",
    "\n",
    "# Write finished mapping file\n",
    "os.makedirs('../temp/', exist_ok = True) # Check for the directory\n",
    "test.to_csv('../temp/TcgaVarImportTest.tsv', sep = '\\t', index = False)  # and create .tsv file in the directory\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
