{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaf4415e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scripts to do:\n",
    "# - Refresher\n",
    "# - Mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3e3cefa",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############################################################################################################\n",
    "##################################### Initialisation ##########################################################\n",
    "###############################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6c934c41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22142"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "from pymongo import MongoClient\n",
    "import bycon\n",
    "import pandas as pd\n",
    "\n",
    "# Read in dataframe from mapping file\n",
    "df = pd.read_csv('../temp/intermediate_mapping_file.csv') ###### Make input later\n",
    "\n",
    "# Prepare mongodb tools\n",
    "client = MongoClient()\n",
    "db = client.progenetix\n",
    "bs = db.biosamples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "dde6aaac",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_ids = df['Tumor_Sample_UUID']\n",
    "mns_ids = df['Matched_Norm_Sample_UUID']\n",
    "case_ids = df['case_id']\n",
    "builds = df['NCBI_Build']\n",
    "chros = df['Chromosome']\n",
    "starts = df['Start_Position']\n",
    "ends = df['End_Position']\n",
    "strands = df['Strand']\n",
    "vcs = df['Variant_Classification']\n",
    "vts = df['Variant_Type']\n",
    "refalls = df['Reference_Allele']\n",
    "tall1s = df['Tumor_Seq_Allele1']\n",
    "tall2s = df['Tumor_Seq_Allele2']\n",
    "hgvscs = df['HGVSc']\n",
    "hgvsps = df['HGVSp']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "4681444c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create varimportTemplate from intfile\n",
    "# - Create varimportTest.tsv with following info\n",
    "#     - variant_id: generic â†’ auto replaced in import\n",
    "#     - callset_id: new; same for all SNV samples\n",
    "#     - biosampled_id: use existing for TCGA sample in DB\n",
    "#     - individual_id: same\n",
    "\n",
    "### IDs and stuff\n",
    "# biosamples id = pgxbs-.... (e.g. pgxbs-kftvi6e8)\n",
    "# individual id = pgxind-... (e.g. pgxind-kftx3nme)\n",
    "# External references:\n",
    "#     case_id = pgx:TCGA-(UUID)\n",
    "#     sample_id = \"\"\n",
    "#     submitter_id = \"\"\n",
    "# Info:\n",
    "#   callset_ids = ['pgxcs-...'] (e.g. pgxcs-kftw4ixr)\n",
    "\n",
    "\n",
    "headers = ['variant_id', 'callset_id', 'case_id', 'biosample_id', 'individual_id', 'sequence_id', 'start', 'end',\n",
    "           'reference_bases', 'alternate_bases_1', 'alternate_bases_2', 'annotation_derived']\n",
    "\n",
    "import_template = pd.DataFrame(columns = headers)\n",
    "import_template['case_id'] = 'pgx:TCGA-' + df['case_id']\n",
    "import_template['start'] = df['Start_Position']\n",
    "import_template['end'] = df['End_Position']\n",
    "import_template['alternate_bases_1'] = df['Tumor_Seq_Allele1']\n",
    "import_template['alternate_bases_2'] = df['Tumor_Seq_Allele2']\n",
    "import_template['reference_bases'] = df['Reference_Allele']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "b2090adb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10108"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###############################################################################################################\n",
    "##################################### Diagnostics #############################################################\n",
    "###############################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "403d06d6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# bs.count_documents({\"cohorts.label\": \"TCGA samples\"}) # 22142\n",
    "\n",
    "# Iterate through sets and count occurences\n",
    "problems = []\n",
    "for ids in case_ids:\n",
    "    n_hits_c = bs.count_documents({\"external_references.id\":{'$regex': ids},\"biosample_status.id\":\"EFO:0009656\"})\n",
    "    if n_hits_c > 1:\n",
    "        problem = bs.find({\"external_references.id\":{'$regex': ids}})\n",
    "        problems.append(problem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf671265",
   "metadata": {},
   "outputs": [],
   "source": [
    "cs = bs.find({\"external_references.id\":{'$regex': case_ids[5]}})\n",
    "for i in cs:\n",
    "    pprint.pprint(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "01491f1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Single hits: 9775 \n",
      "Multiple hits: 125 \n",
      "No hit: 208 \n",
      "Total: 10108\n"
     ]
    }
   ],
   "source": [
    "single_hit = 0\n",
    "multiple_hits = 0\n",
    "no_hit = 0\n",
    "for i in hits_c:\n",
    "    if i == 1:\n",
    "        single_hit += 1\n",
    "    if i == 0:\n",
    "        no_hit += 1\n",
    "    if i != 0 and i != 1:\n",
    "        multiple_hits += 1\n",
    "\n",
    "print(\"Single hits:\", single_hit, \"\\nMultiple hits:\", multiple_hits,\n",
    "      \"\\nNo hit:\", no_hit, \"\\nTotal:\", len(hits_c))\n",
    "# Single hits: 9775 \n",
    "# Multiple hits: 125 \n",
    "# No hit: 208 \n",
    "# Total: 10108\n",
    "\n",
    "# Legacy ids = sample id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b3b07f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Findings!\n",
    "# The sample ids have hierarchy:\n",
    "# Case UUID: becedbfd-b2aa-4dde-b7f4-29e6f59ec32c in data bank\n",
    "# Sample UUID: f6ee8148-9fb7-404e-b951-d1af8d06974f not in data bank\n",
    "# Belongs to sample TCGA-E2-A1IK-01A\n",
    "# Sample UUID: 8577ac01-1274-4bd5-ab04-380eaa78d95b in data bank\n",
    "# Belongs to aliquot TCGA-E2-A1IK-01A-11D-A17G-09\n",
    "# TCGA-E2-A1IK-01A\n",
    "#     Portions\n",
    "#     TCGA-E2-A1IK-01A-11\n",
    "#         Analytes\n",
    "#         TCGA-E2-A1IK-01A-11D\n",
    "#             Aliquots\n",
    "#             TCGA-E2-A1IK-01A-11D-A141-01\n",
    "#             TCGA-E2-A1IK-01A-11D-A17G-09\n",
    "#             TCGA-E2-A1IK-01A-11D-A140-02\n",
    "#             TCGA-E2-A1IK-01A-11D-A142-09\n",
    "#             TCGA-E2-A1IK-01A-11D-A145-05\n",
    "\n",
    "# Solution: find sample UUID of aliquot sample UUID\n",
    "\n",
    "\n",
    "# COSMIC is in TCGA\n",
    "# 118 - COSMIC\tOverlapping COSMIC variants\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "6b29b546",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'generate_id' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[70], line 25\u001b[0m\n\u001b[1;32m     22\u001b[0m         variants\u001b[38;5;241m.\u001b[39mappend(v)\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m variants, v_e\n\u001b[0;32m---> 25\u001b[0m \u001b[43mgenerate_id\u001b[49m()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'generate_id' is not defined"
     ]
    }
   ],
   "source": [
    "# bycon generate_id\n",
    "\n",
    "from helpers import generate_id\n",
    "\n",
    "def variants_from_revish(bs_id, cs_id, technique, iscn, byc):\n",
    "\n",
    "    v_s, v_e = deparse_ISCN_to_variants(iscn, byc)\n",
    "    variants = []\n",
    "\n",
    "    # the id here is a placeholder since we now use a stringified version of the\n",
    "    # MongoDB ObjectId w/ `pgxvar-` prepend\n",
    "    for v in v_s:\n",
    "\n",
    "        v.update({\n",
    "            \"id\": generate_id(\"pgxvar\"),\n",
    "            \"variant_internal_id\": variant_create_digest(v, byc),\n",
    "            \"biosample_id\": bs_id,\n",
    "            \"callset_id\": cs_id,\n",
    "            \"updated\": datetime.datetime.now().isoformat()\n",
    "        })\n",
    "\n",
    "        variants.append(v)\n",
    "\n",
    "    return variants, v_e\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3f48a1d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [-t TEST] [-i INPUTFILE] [-s SURVIVALFILE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /Users/kayvongrunigen/Library/Jupyter/runtime/kernel-ce3c82a0-b671-426e-b557-5e0b2a949c95.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/mambaforge/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3468: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "# mapTCGAids.py\n",
    "#!/usr/local/bin/python3\n",
    "\n",
    "from pymongo import MongoClient\n",
    "import argparse\n",
    "import json\n",
    "import re\n",
    "from progress.bar import Bar\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "\n",
    "def _get_args():\n",
    "\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument(\"-t\", \"--test\", help=\"test setting\")\n",
    "    parser.add_argument('-i', '--inputfile', help='a custom file to specify input data')\n",
    "    parser.add_argument('-s', '--survivalfile', help='...')\n",
    "\n",
    "    return parser.parse_args()\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def main():\n",
    "\n",
    "    args = _get_args()\n",
    "\n",
    "    ds_id = 'progenetix'\n",
    "    \n",
    "    mapTCGAids(args, ds_id)\n",
    "    addTCGAsurvival(args, ds_id)\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def mapTCGAids(args, ds_id): #re-construct the external descriptions, and add submitter and case ids  of TCGA sampels\n",
    "\n",
    "    if not args.inputfile:\n",
    "        print( \"Not updating ids - you have to provide an inputfile -i\")\n",
    "        return\n",
    "\n",
    "    data_client = MongoClient('localhost', 27017)\n",
    "    data_db = data_client[ ds_id ]\n",
    "    bios_coll = data_db[ \"biosamples\" ]\n",
    "    ind_coll = data_db[ \"individuals\" ]\n",
    "\n",
    "    ind_ids = {}\n",
    "\n",
    "    fh=open(args.inputfile, 'r')\n",
    "    idlist = [line.rstrip() for line in fh.readlines()[1:]]\n",
    "    fh.close()\n",
    "\n",
    "    no = len(idlist)\n",
    "    if not args.test:\n",
    "        bar = Bar(\"{} samples\".format(no), max = no, suffix='%(percent)d%%'+\" of \"+str(no) )\n",
    "    else:\n",
    "        idlist = idlist[0:20]\n",
    "\n",
    "    for line in idlist:\n",
    "        tcga_s_id, tcga_c_id, tcga_submitter_id  = line.rstrip().split('\\t')\n",
    "\n",
    "        q = {'info.legacy_ids': tcga_s_id}\n",
    "        b = bios_coll.find_one(q)\n",
    "\n",
    "        i_e_r = [\n",
    "            {\"id\": \"pgx:TCGA-\"+tcga_c_id, \"label\": \"TCGA case_id\"},\n",
    "            {\"id\": \"pgx:TCGA-\"+tcga_submitter_id, \"label\": \"TCGA submitter_id\"}\n",
    "        ]\n",
    "\n",
    "        if tcga_c_id in ind_ids:\n",
    "            ind_ids[tcga_c_id][\"biosample_ids\"].append(b[\"id\"])\n",
    "        else:\n",
    "            ind_ids.update({\n",
    "                tcga_c_id: {\n",
    "                    \"external_references\": i_e_r,\n",
    "                    \"biosample_ids\": [b[\"id\"]]\n",
    "                }\n",
    "            })\n",
    "\n",
    "        b_e_r = i_e_r.copy()\n",
    "        b_e_r.append({\"id\": \"pgx:TCGA-\"+tcga_s_id, \"label\": \"TCGA sample_id\"})\n",
    "\n",
    "        for e in b[\"external_references\"]:\n",
    "            if not \"TCGA\" in e[\"id\"]: # <= important to keep other ers\n",
    "                b_e_r.append(e)\n",
    "            if re.match(r'^(?:pgx:)?TCGA\\-\\w+?$', e[\"id\"]):\n",
    "                project = re.sub('TCGA-', '', e[\"id\"])\n",
    "                project = re.sub('pgx:', '', project)\n",
    "                b_e_r.append({\"id\": \"pgx:TCGA-\"+project, \"label\": \"TCGA \"+project+\" project\"})\n",
    "\n",
    "        if args.test:\n",
    "            print(tcga_c_id, ind_ids[tcga_c_id][\"biosample_ids\"])\n",
    "        else:\n",
    "            bar.next()\n",
    "            bios_coll.update_one({\"_id\":b[\"_id\"]}, {\"$set\": {\"external_references\": b_e_r}})\n",
    "\n",
    "\n",
    "    if not args.test:\n",
    "        bar.finish()\n",
    "\n",
    "    ############################################################################\n",
    "\n",
    "    no = len(ind_ids.keys())\n",
    "    if not args.test:\n",
    "        bar = Bar(\"{} individuals\".format(no), max = no, suffix='%(percent)d%%'+\" of \"+str(no) )\n",
    "\n",
    "    for i_id, i_data in ind_ids.items():\n",
    "\n",
    "        i_e_r = i_data[\"external_references\"]\n",
    "        i_e_r_ids = set()\n",
    "\n",
    "        ind_ids = bios_coll.distinct(\"individual_id\", {\"id\":{\"$in\":i_data[\"biosample_ids\"]}})\n",
    "\n",
    "        i_s = ind_coll.find({\"id\":{\"$in\":ind_ids}})\n",
    "        for i in i_s:\n",
    "            for e in i[\"external_references\"]:\n",
    "                if not \"TCGA\" in e[\"id\"]: # <= important to keep other ers\n",
    "                    if not e[\"id\"] in i_e_r_ids:\n",
    "                        i_e_r_ids.add(e[\"id\"])\n",
    "                        i_e_r.append(e)\n",
    "\n",
    "        ind_coll.update_many({\"id\":{\"$in\":ind_ids}}, {\"$set\":{\"external_references\":i_e_r}})\n",
    "\n",
    "        if len(ind_ids) > 1:\n",
    "            ind_ids = ind_ids[1:]\n",
    "\n",
    "            print(\"deleting {} of {} individuals\".format(len(ind_ids), len(i_data[\"biosample_ids\"])))\n",
    "\n",
    "            ind_coll.delete_many({\"id\":{\"$in\":ind_ids}})\n",
    "\n",
    "\n",
    "# def mapTCGAsurvival(clinical,output):# map survival information to TCGA samples\n",
    "#     with open(output + '.csv', 'w') as f:\n",
    "#         f.write('case_id')\n",
    "#         f.write('\\t')\n",
    "#         f.write('vital_status')\n",
    "#         f.write('\\n')\n",
    "#     client = MongoClient('localhost', 27017)\n",
    "#     #db = client[database]\n",
    "#     file = open(clinical + '.json', )\n",
    "#     data = json.load(file)\n",
    "#     for i in data:\n",
    "#         with open(output + '.csv', 'a+') as f:\n",
    "#             f.write(i['case_id'])\n",
    "#             f.write('\\t')\n",
    "#             f.write(i['demographic']['vital_status'])\n",
    "#             f.write('\\n')\n",
    "\n",
    "def addTCGAsurvival(args, ds_id):# add survival information for cancer TCGA samples\n",
    "\n",
    "    if not args.survivalfile:\n",
    "        print( \"You have to provide an survivalfile -s\")\n",
    "        return\n",
    "\n",
    "    data_client = MongoClient('localhost', 27017)\n",
    "    data_db = data_client[ ds_id ]\n",
    "    bios_coll = data_db[ \"biosamples\" ]\n",
    "    ind_coll = data_db[ \"individuals\" ]\n",
    "\n",
    "    fh=open(args.survivalfile, 'r')\n",
    "    survlist = [line.rstrip() for line in fh.readlines()[1:]]\n",
    "    fh.close()\n",
    "\n",
    "    no = len(survlist)\n",
    "    if not args.test:\n",
    "        bar = Bar(\"{} samples\".format(no), max = no, suffix='%(percent)d%%'+\" of \"+str(no) )\n",
    "    else:\n",
    "        survlist = survlist[0:20]\n",
    "\n",
    "    for line in survlist:\n",
    "        temp = line.split('\\t')\n",
    "        e_r_s_id = \"pgx:TCGA-\"+temp[0]\n",
    "        surv = temp[1].lower()\n",
    "        bs_q = {'external_references.id': e_r_s_id,'cohorts.id':'pgx:cohort-TCGAcancers'}\n",
    "        if args.test:\n",
    "            print(e_r_s_id, surv)\n",
    "        else:\n",
    "            bar.next()\n",
    "            bios_coll.update_many(bs_q, {'$set': {'info.survival_status': surv}})\n",
    "\n",
    "    if not args.test:\n",
    "        bar.finish()\n",
    "        \n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()\n",
    "\n",
    "#mapTCGAids('TCGA_ids.txt','progenetix')\n",
    "#addTCGAsurvival('TCGAcase_survival','progenetixCopy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67314a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tcga_mapper.py\n",
    "#!/usr/local/bin/python3\n",
    "\n",
    "from pymongo import MongoClient\n",
    "import json\n",
    "import re\n",
    "import json\n",
    "import argparse\n",
    "from os import path\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def _get_args():\n",
    "\n",
    "    parser = argparse.ArgumentParser()\n",
    "\n",
    "    parser.add_argument(\"-d\", \"--datasetid\", help=\"dataset, usually progenetix\")\n",
    "    parser.add_argument('-t', '--tcgafile', help='TCGA json data file')\n",
    "    parser.add_argument('-i', '--idfile', help='idfile')\n",
    "    parser.add_argument('-u', '--update', help='update modus')\n",
    "\n",
    "    return parser.parse_args()\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def main():\n",
    "\n",
    "\targs = _get_args()\n",
    "\n",
    "\tupdate_TCGA_data(args)\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def update_TCGA_data(args):\n",
    "\n",
    "\tif not args.datasetid:\n",
    "\t\tprint(\"no datasetid was specified with -d\")\n",
    "\t\texit()\n",
    "\n",
    "\tif not args.idfile:\n",
    "\t\tprint(\"no id file was specified with -i\")\n",
    "\t\texit()\n",
    "\n",
    "\tif not args.tcgafile:\n",
    "\t\tprint(\"no tcga file was specified with -t\")\n",
    "\t\texit()\n",
    "\n",
    "\tif not args.update:\n",
    "\t\tprint('no update will be performed - activate with \"-u 1\"')\n",
    "\n",
    "\tmongo_client = MongoClient()\n",
    "\tdata_db = mongo_client[args.datasetid]\n",
    "\tbios_coll = data_db[\"biosamples\"]\n",
    "\tind_coll = data_db[\"individuals\"]\n",
    "\n",
    "\tf = open(args.idfile, 'r+')\n",
    "\tid_lines = [line for line in f.readlines()]\n",
    "\tf.close()\n",
    "\n",
    "\ttcga_ids = {}\n",
    "\n",
    "\tfor id_l in id_lines:\n",
    "\t\tid_l = id_l.rstrip()\n",
    "\t\tsample_id, case_id, submitter_id = re.split(\"\\t\", id_l)\n",
    "\t\ttcga_ids.update({sample_id:{\"sample_id\": sample_id, \"case_id\": case_id, \"submitter_id\": submitter_id }})\n",
    "\n",
    "\twith open(args.tcgafile, 'r') as c_f:\n",
    "\t\tc = c_f.read()\n",
    "\t\tc_l = json.loads(c)\n",
    "\n",
    "\ttcga_clin = {}\n",
    "\n",
    "\tfor tcga_c in c_l:\n",
    "\t\ttcga_clin.update({ tcga_c[\"case_id\"]: tcga_c })\n",
    "\n",
    "\tfor s_id, s_id_o in tcga_ids.items():\n",
    "\n",
    "\t\ts = bios_coll.find_one({\"info.legacy_id\": s_id })\n",
    "\t\tif not s:\n",
    "\t\t\tprint('!!! no biosample for \"{}\"'.format(s_id))\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tc_id = s_id_o[\"case_id\"]\n",
    "\n",
    "\t\tif not c_id in tcga_clin:\n",
    "\t\t\tprint('!!! no tcga data for case_id \"{}\"'.format(c_id))\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tclin = tcga_clin[c_id]\n",
    "\n",
    "\t\tbios_update = {\n",
    "\t\t\t\"external_references\": [],\n",
    "\t\t\t\"info\": s.get(\"info\", {})     # <=== important!\n",
    "\t\t}\n",
    "\n",
    "\t\t# references\n",
    "\t\t# first collecting all \"non-TCGA ones\" for keeping, then adding the new TCGA ones\n",
    "\t\tfor pgx_e_r in s[\"external_references\"]:\n",
    "\t\t\tif \"TCGA\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\tif not \"project\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\t\t# skipping previous TCGA labels _except_ the TCGA-XXX for projects\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\tpgx_e_r.pop(\"description\", None)\n",
    "\t\t\tbios_update[\"external_references\"].append(pgx_e_r)\n",
    "\n",
    "\t\t\tfor id_t in s_id_o.keys():\n",
    "\t\t\t\tn_id = s_id_o[id_t]\n",
    "\t\t\t\tif not \"TCGA-\" in n_id:\n",
    "\t\t\t\t\tn_id = \"TCGA-\"+s_id_o[id_t]\n",
    "\t\t\t\tid_obj = { \"id\": n_id, \"label\": \"TCGA \"+id_t }\n",
    "\t\t\t\tbios_update[\"external_references\"].append(id_obj)\n",
    "\n",
    "\t\t# clinical\n",
    "\t\tf_u_s = clin[\"demographic\"].get(\"vital_status\", \"\").lower()\n",
    "\t\tif \"dead\" in f_u_s:\n",
    "\t\t\tbios_update.update({\"followup_state\":{\"id\":\"EFO:0030049\", \"label\":\"dead (follow-up status)\"}})\n",
    "\t\t\tbios_update[\"info\"].update({\"survival_status\":\"dead\"})\n",
    "\t\telif \"alive\" in f_u_s:\n",
    "\t\t\tbios_update.update({\"followup_state\":{\"id\":\"EFO:0030041\", \"label\":\"alive (follow-up status)\"}})\n",
    "\t\t\tbios_update[\"info\"].update({\"survival_status\":\"alive\"})\n",
    "        #Ziying update => since [\"demographic\"][\"days_to_death\"] should be correct for dead samples\n",
    "        if clin[\"demographic\"][\"days_to_death\"]:\n",
    "            f_u_d = clin[\"demographic\"][\"days_to_death\"]\n",
    "        else:\n",
    "            f_u_d = clin[\"diagnoses\"][0][\"days_to_last_follow_up\"]\n",
    "\t\t#f_u_d = clin[\"diagnoses\"][0][\"days_to_last_follow_up\"]\n",
    "\t\tif isinstance(f_u_d, int):\n",
    "\t\t\tf_u_m = round(f_u_d/30.5)\n",
    "\t\t\tbios_update[\"info\"].update({\"followup_months\":f_u_m})\n",
    "\n",
    "\t\ta_d = clin[\"diagnoses\"][0][\"age_at_diagnosis\"]\n",
    "\t\tif isinstance(a_d, int):\n",
    "\t\t\ta = _d2iso(a_d)\n",
    "\t\t\tbios_update.update({\"individual_age_at_collection\":\"{}\".format(a)})\n",
    "\n",
    "\t\tif args.update:\n",
    "\t\t\tbios_coll.update_one({\"_id\": s[\"_id\"] }, { '$set': bios_update } )\n",
    "\t\telse:\n",
    "\t\t\tprint(bios_update)\n",
    "\n",
    "\t\t########################################################################\n",
    "\t\t# individuals\n",
    "\t\t########################################################################\n",
    "\n",
    "\t\ti = ind_coll.find_one({\"id\": s[\"individual_id\"] })\n",
    "\t\tif not s:\n",
    "\t\t\tprint('!!! no individual for \"{}\"'.format(s[\"individual_id\"]))\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tind_update = {\n",
    "\t\t\t\"external_references\": [],\n",
    "\t\t\t\"info\": i.get(\"info\", {}),\n",
    "\t\t}\n",
    "\n",
    "\t\t# references\n",
    "\t\tfor pgx_e_r in s[\"external_references\"]:\n",
    "\t\t\tif \"TCGA\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\tif \"biosample\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\t\tif \"TCGA collection\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\tpgx_e_r.pop(\"description\", None)\n",
    "\t\t\tbios_update[\"external_references\"].append(pgx_e_r)\n",
    "\t\t\tind_update[\"external_references\"].append(pgx_e_r)\n",
    "\n",
    "\t\t\tfor id_t in s_id_o.keys():\n",
    "\t\t\t\tid_obj = { \"id\": \"TCGA-\"+s_id_o[id_t], \"label\": \"TCGA \"+id_t }\n",
    "\t\t\t\tbios_update[\"external_references\"].append(id_obj)\n",
    "\t\t\t\tind_update[\"external_references\"].append(id_obj)\n",
    "\n",
    "\t\tsex = clin[\"demographic\"].get(\"gender\", \"\")\n",
    "\t\tif \"female\" in sex:\n",
    "\t\t\tind_update.update({\"sex\":{\"id\":\"PATO:0020002\", \"label\":\"female genotypic sex\"}})\n",
    "\t\telif \"male\" in sex:\n",
    "\t\t\tind_update.update({\"sex\":{\"id\":\"PATO:0020001\", \"label\":\"male genotypic sex\"}})\n",
    "\n",
    "\t\tif args.update:\n",
    "\t\t\tind_coll.update_one({\"_id\": i[\"_id\"] }, { '$set': ind_update } )\n",
    "\t\telse:\n",
    "\t\t\tprint(ind_update)\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def _d2iso(d):\n",
    "\n",
    "\ty = int(d / 365.25)\n",
    "\tm = int((d % 365.25) / 30.5)\n",
    "\td = int(d - (y * 365.25 + m * 30.5))\n",
    "\n",
    "\treturn \"P{}Y{}M{}D\".format(y,m,d)\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main(  )\n",
    "\n",
    "    \n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "    \n",
    "# tcga_mapperZY.py    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    #!/usr/local/bin/python3\n",
    "\n",
    "from pymongo import MongoClient\n",
    "import json\n",
    "import re\n",
    "import json\n",
    "import argparse\n",
    "from os import path\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def _get_args():\n",
    "\n",
    "    parser = argparse.ArgumentParser()\n",
    "\n",
    "    parser.add_argument(\"-d\", \"--datasetid\", help=\"dataset, usually progenetix\")\n",
    "    parser.add_argument('-t', '--tcgafile', help='TCGA json data file')\n",
    "    parser.add_argument('-i', '--idfile', help='idfile')\n",
    "    parser.add_argument('-u', '--update', help='update modus')\n",
    "\n",
    "    return parser.parse_args()\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def main():\n",
    "\n",
    "\targs = _get_args()\n",
    "\n",
    "\tupdate_TCGA_data(args)\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def update_TCGA_data(args):\n",
    "\n",
    "\tif not args.datasetid:\n",
    "\t\tprint(\"no datasetid was specified with -d\")\n",
    "\t\texit()\n",
    "\n",
    "\tif not args.idfile:\n",
    "\t\tprint(\"no id file was specified with -i\")\n",
    "\t\texit()\n",
    "\n",
    "\tif not args.tcgafile:\n",
    "\t\tprint(\"no tcga file was specified with -t\")\n",
    "\t\texit()\n",
    "\n",
    "\tif not args.update:\n",
    "\t\tprint('no update will be performed - activate with \"-u 1\"')\n",
    "\n",
    "\tmongo_client = MongoClient()\n",
    "\tdata_db = mongo_client[args.datasetid]\n",
    "\tbios_coll = data_db[\"biosamples\"]\n",
    "\tind_coll = data_db[\"individuals\"]\n",
    "\n",
    "\tf = open(args.idfile, 'r+')\n",
    "\tid_lines = [line for line in f.readlines()]\n",
    "\tf.close()\n",
    "\n",
    "\ttcga_ids = {}\n",
    "\n",
    "\tfor id_l in id_lines:\n",
    "\t\tid_l = id_l.rstrip()\n",
    "\t\tsample_id, case_id, submitter_id = re.split(\"\\t\", id_l)\n",
    "\t\ttcga_ids.update({sample_id:{\"sample_id\": sample_id, \"case_id\": case_id, \"submitter_id\": submitter_id }})\n",
    "\n",
    "\twith open(args.tcgafile, 'r') as c_f:\n",
    "\t\tc = c_f.read()\n",
    "\t\tc_l = json.loads(c)\n",
    "\n",
    "\ttcga_clin = {}\n",
    "\n",
    "\tfor tcga_c in c_l:\n",
    "\t\ttcga_clin.update({ tcga_c[\"case_id\"]: tcga_c })\n",
    "\n",
    "\tfor s_id, s_id_o in tcga_ids.items():\n",
    "\n",
    "\t\ts = bios_coll.find_one({\"info.legacy_id\": s_id })\n",
    "\t\tif not s:\n",
    "\t\t\tprint('!!! no biosample for \"{}\"'.format(s_id))\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tc_id = s_id_o[\"case_id\"]\n",
    "\n",
    "\t\tif not c_id in tcga_clin:\n",
    "\t\t\tprint('!!! no tcga data for case_id \"{}\"'.format(c_id))\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tclin = tcga_clin[c_id]\n",
    "\n",
    "\t\tbios_update = {\n",
    "\t\t\t#\"external_references\": [],\n",
    "\t\t\t\"info\": s.get(\"info\", {})     # <=== important!\n",
    "\t\t}\n",
    "\n",
    "\t\t# references\n",
    "\t\t#first collecting all \"non-TCGA ones\" for keeping, then adding the new TCGA ones\n",
    "\t\t# for pgx_e_r in s[\"external_references\"]:\n",
    "\t\t# \tif \"TCGA\" in pgx_e_r[\"label\"]:\n",
    "\t\t# \t\tif not \"project\" in pgx_e_r[\"label\"]:\n",
    "\t\t# \t\t\t# skipping previous TCGA labels _except_ the TCGA-XXX for projects\n",
    "\t\t# \t\t\tcontinue\n",
    "\t\t# \tpgx_e_r.pop(\"description\", None)\n",
    "\t\t# \tbios_update[\"external_references\"].append(pgx_e_r)\n",
    "\t\t#\n",
    "\t\t# \tfor id_t in s_id_o.keys():\n",
    "\t\t# \t\tn_id = s_id_o[id_t]\n",
    "\t\t# \t\tif not \"TCGA-\" in n_id:\n",
    "\t\t# \t\t\tn_id = \"TCGA-\"+s_id_o[id_t]\n",
    "\t\t# \t\tid_obj = { \"id\": n_id, \"label\": \"TCGA \"+id_t }\n",
    "\t\t# \t\tbios_update[\"external_references\"].append(id_obj)\n",
    "\n",
    "\t\t# clinical\n",
    "\t\tf_u_s = clin[\"demographic\"].get(\"vital_status\", \"\").lower()\n",
    "\t\tif \"dead\" in f_u_s:\n",
    "\t\t\tbios_update.update({\"followup_state\":{\"id\":\"EFO:0030049\", \"label\":\"dead (follow-up status)\"}})\n",
    "\t\t\tbios_update[\"info\"].update({\"survival_status\":\"dead\"})\n",
    "\t\telif \"alive\" in f_u_s:\n",
    "\t\t\tbios_update.update({\"followup_state\":{\"id\":\"EFO:0030041\", \"label\":\"alive (follow-up status)\"}})\n",
    "\t\t\tbios_update[\"info\"].update({\"survival_status\":\"alive\"})\n",
    "\t\tif \"days_to_death\" in clin[\"demographic\"]:\n",
    "\t\t\tf_u_d = clin[\"demographic\"][\"days_to_death\"]\n",
    "\t\telse:\n",
    "\t\t\tf_u_d = clin[\"diagnoses\"][0][\"days_to_last_follow_up\"]\n",
    "\n",
    "\t\t#update info clinical\n",
    "\t\tif \"ajcc_clinical_stage\" in clin[\"diagnoses\"][0]:\n",
    "\t\t\tbios_update[\"info\"].update({\"tumor_stage\":clin[\"diagnoses\"][0][\"ajcc_clinical_stage\"].split(' ')[1]})\n",
    "\t\tif \"tumor_grade\" in clin[\"diagnoses\"][0]:\n",
    "\t\t\tbios_update[\"info\"].update({\"tumor_grade\": clin[\"diagnoses\"][0][\"tumor_grade\"]})\n",
    "\t\tt=n=m=''\n",
    "\t\tif \"ajcc_pathologic_t\" in clin[\"diagnoses\"][0]:\n",
    "\t\t\tt = clin[\"diagnoses\"][0][\"ajcc_pathologic_t\"]\n",
    "\t\tif \"ajcc_pathologic_n\" in clin[\"diagnoses\"][0]:\n",
    "\t\t\tt = clin[\"diagnoses\"][0][\"ajcc_pathologic_n\"]\n",
    "\t\tif \"ajcc_pathologic_m\" in clin[\"diagnoses\"][0]:\n",
    "\t\t\tt = clin[\"diagnoses\"][0][\"ajcc_pathologic_m\"]\n",
    "\t\ttnm = t+n+m\n",
    "\t\tif tnm:\n",
    "\t\t\tbios_update[\"info\"].update({\"tnm\": tnm})\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\t\tif isinstance(f_u_d, int):\n",
    "\t\t\tf_u_m = round(f_u_d/30.5)\n",
    "\t\t\tbios_update[\"info\"].update({\"followup_months\":f_u_m})\n",
    "\n",
    "\t\ta_d = clin[\"diagnoses\"][0][\"age_at_diagnosis\"]\n",
    "\t\tif isinstance(a_d, int):\n",
    "\t\t\ta = _d2iso(a_d)\n",
    "\t\t\tbios_update.update({\"individual_age_at_collection\":\"{}\".format(a)})\n",
    "\n",
    "\t\tif args.update:\n",
    "\t\t\tbios_coll.update_one({\"_id\": s[\"_id\"] }, { '$set': bios_update } )\n",
    "\t\telse:\n",
    "\t\t\tprint(bios_update)\n",
    "\n",
    "\t\t########################################################################\n",
    "\t\t# individuals\n",
    "\t\t########################################################################\n",
    "\n",
    "\t\ti = ind_coll.find_one({\"id\": s[\"individual_id\"] })\n",
    "\t\tif not s:\n",
    "\t\t\tprint('!!! no individual for \"{}\"'.format(s[\"individual_id\"]))\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tind_update = {\n",
    "\t\t\t\"external_references\": [],\n",
    "\t\t\t\"info\": i.get(\"info\", {}),\n",
    "\t\t}\n",
    "\n",
    "\t\t# references\n",
    "\t\tfor pgx_e_r in s[\"external_references\"]:\n",
    "\t\t\tif \"TCGA\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\tif \"biosample\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\t\tif \"TCGA collection\" in pgx_e_r[\"label\"]:\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\tpgx_e_r.pop(\"description\", None)\n",
    "\t\t\t#bios_update[\"external_references\"].append(pgx_e_r)\n",
    "\t\t\tind_update[\"external_references\"].append(pgx_e_r)\n",
    "\n",
    "\t\t\tfor id_t in s_id_o.keys():\n",
    "\t\t\t\tid_obj = { \"id\": \"TCGA-\"+s_id_o[id_t], \"label\": \"TCGA \"+id_t }\n",
    "\t\t\t\t#bios_update[\"external_references\"].append(id_obj)\n",
    "\t\t\t\tind_update[\"external_references\"].append(id_obj)\n",
    "\n",
    "\t\tsex = clin[\"demographic\"].get(\"gender\", \"\")\n",
    "\t\tif \"female\" in sex:\n",
    "\t\t\tind_update.update({\"sex\":{\"id\":\"PATO:0020002\", \"label\":\"female genotypic sex\"}})\n",
    "\t\telif \"male\" in sex:\n",
    "\t\t\tind_update.update({\"sex\":{\"id\":\"PATO:0020001\", \"label\":\"male genotypic sex\"}})\n",
    "\n",
    "\t\t#update diseases in individuals\n",
    "\n",
    "\n",
    "\t\tif args.update:\n",
    "\t\t\tind_coll.update_one({\"_id\": i[\"_id\"] }, { '$set': ind_update } )\n",
    "\t\telse:\n",
    "\t\t\tprint(ind_update)\n",
    "\n",
    "################################################################################\n",
    "\n",
    "def _d2iso(d):\n",
    "\n",
    "\ty = int(d / 365.25)\n",
    "\tm = int((d % 365.25) / 30.5)\n",
    "\td = int(d - (y * 365.25 + m * 30.5))\n",
    "\n",
    "\treturn \"P{}Y{}M{}D\".format(y,m,d)\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main(  )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
